notebook/gateway/handlers.py

```diff
@@ -0,0 +1,354 @@
+            else:
+                # This is now an async operation
+                model = yield gen.maybe_future(km.kernel_model(kernel_id))
+                self.write(json.dumps(model))
+        self.finish()
+
+# -----------------------------------------------------------------------------
+# kernel spec handlers
+# -----------------------------------------------------------------------------
+
+
+class MainKernelSpecHandler(APIHandler):
+    @web.authenticated
+    @gen.coroutine
+    def get(self):
+        ksm = self.kernel_spec_manager
+        try:
+            kernel_specs = yield gen.maybe_future(ksm.list_kernel_specs())
+            # TODO: Remove resources until we support them
+            for name, spec in kernel_specs['kernelspecs'].items():
+                spec['resources'] = {}
+            self.set_header("Content-Type", 'application/json')
+            self.write(json.dumps(kernel_specs))
+
+        # Trap a set of common exceptions so that we can inform the user that their Gateway url is incorrect
+        # or the server is not running.
+        # NOTE: We do this here since this handler is called during the Notebook's startup and subsequent refreshes
+        # of the tree view.
+        except ConnectionRefusedError:
+            self.log.error("Connection refused from Gateway server url '{}'.  "
+                         "Check to be sure the Gateway instance is running.".format(Gateway.instance().url))
+        except HTTPTimeoutError:
+            # This can occur if the host is valid (e.g., foo.com) but there's nothing there.
+            self.log.error("Timeout error attempting to connect to Gateway server url '{}'.  "
+                         "Ensure gateway url is valid and the Gateway instance is running.".format(Gateway.instance().url))
+        except gaierror as e:
+            self.log.error("The Gateway server specified in the gateway_url '{}' doesn't appear to be valid.  "
+                         "Ensure gateway url is valid and the Gateway instance is running.".format(Gateway.instance().url))
+
+        self.finish()
+
+
+class KernelSpecHandler(APIHandler):
+    @web.authenticated
+    @gen.coroutine
+    def get(self, kernel_name):
+        ksm = self.kernel_spec_manager
+        kernel_spec = yield ksm.get_kernel_spec(kernel_name)
+        if kernel_spec is None:
+            raise web.HTTPError(404, u'Kernel spec %s not found' % kernel_name)
+        # TODO: Remove resources until we support them
+        kernel_spec['resources'] = {}
+        self.set_header("Content-Type", 'application/json')
+        self.finish(json.dumps(kernel_spec))
+
+# -----------------------------------------------------------------------------
+# URL to handler mappings
+# -----------------------------------------------------------------------------
+
+
+from ..services.kernels.handlers import _kernel_id_regex, _kernel_action_regex
+from ..services.kernelspecs.handlers import kernel_name_regex
+
+default_handlers = [
```

Comment:
There appears to be a lot of duplication of handlers here. Ideally, there would be ~no duplication of Handlers, only the Manager implementations, but I think the channels URL at least needs special handling. It doesn't appear to me that the kernel listing, action, or kernel spec handlers need to be replaced, though.

Reply:
Good point - that makes sense.  Taking a quick look, it appears there are some methods that have been added to Notebook since NB2KG was first implemented.  I should be able to create appropriate methods (that hit the gateway or equivalent).  I'll move this back to WIP again.

Question...  How would you recommend the handler load for path `r"/api/kernels/%s/channels" % _kernel_id_regex` be handled, since use of the gateway will still require the NB kernel handlers - where its own channels handler is also registered?  Does the order of the handlers list matter and, if so, is it a _first match wins_ behavior?  If handlers are unordered, then I'll need to conditionally perform replacement within the handlers list in `init_handlers()` to get the appropriate channels handler registered.